fdbrpc/HTTPServer.actor.cpp

```diff
@@ -0,0 +1,289 @@
+                                   FlowMutex* mutex) {
+	state Reference<HTTP::OutgoingResponse> response = makeReference<HTTP::OutgoingResponse>();
+	state UnsentPacketQueue content;
+	response->data.content = &content;
+	response->data.contentLen = 0;
+
+	try {
+		wait(readRequestDone);
+		wait(callback(req, response));
+	} catch (Error& e) {
+		if (e.code() == error_code_operation_cancelled) {
+			throw e;
+		}
+		// FIXME: other errors?
+		if (e.code() == error_code_http_request_failed || e.code() == error_code_http_bad_response ||
+		    e.code() == error_code_connection_failed) {
+			TraceEvent(SevWarn, "HTTPServerConnHandlerInternalError").errorUnsuppressed(e);
+			// reset to empty error response
+			response->reset();
+			response->code = 500;
+		} else {
+			TraceEvent(SevWarn, "HTTPServerConnHandlerUnexpectedError").errorUnsuppressed(e);
+			throw e;
+		}
+	}
+	// take out response mutex to ensure no parallel writers to response connection
+	// FIXME: is this necessary? I think it is
+	state FlowMutex::Lock lock = wait(mutex->take());
+	try {
+		wait(response->write(conn));
+	} catch (Error& e) {
+		lock.release();
+		if (e.code() == error_code_connection_failed) {
+			// connection back to client failed, end. They will retry if they still need the response.
+			TraceEvent("HTTPServerConnHandlerResponseError").errorUnsuppressed(e);
+			return Void();
+		}
+		TraceEvent("HTTPServerConnHandlerResponseUnexpectedError").errorUnsuppressed(e);
+		throw e;
+	}
+	lock.release();
+	return Void();
+}
+
+ACTOR Future<Void> connectionHandler(Reference<HTTP::SimServerContext> server,
+                                     Reference<IConnection> conn,
+                                     HTTP::ServerCallback callback) {
+	try {
+		// TODO do we actually have multiple requests on a connection? how does this work
+		state FlowMutex responseMutex;
+		state Future<Void> readPrevRequest = Future<Void>(Void());
+		wait(conn->acceptHandshake());
+		loop {
+			wait(readPrevRequest);
+			wait(delay(0));
+			wait(conn->onReadable());
+			state Reference<HTTP::IncomingRequest> req = makeReference<HTTP::IncomingRequest>();
+			readPrevRequest = req->read(conn, false);
+			server->actors.add(callbackHandler(conn, readPrevRequest, callback, req, &responseMutex));
+		}
+	} catch (Error& e) {
+		if (e.code() != error_code_actor_cancelled) {
+			TraceEvent("HTTPConnectionError", server->dbgid)
+			    .errorUnsuppressed(e)
+			    .suppressFor(1.0)
+			    .detail("ConnID", conn->getDebugID())
+			    .detail("FromAddress", conn->getPeerAddress());
+		}
+		conn->close();
+	}
+	return Void();
+}
+
+ACTOR Future<Void> listenActor(Reference<HTTP::SimServerContext> server,
+                               HTTP::ServerCallback callback,
+                               NetworkAddress addr,
+                               Reference<IListener> listener) {
+	// TODO: reduce trace verbosity and/or SevDebug?
```

Comment:
an alternate would be introduce HTTL_SERVER_LOG_LEVEL  and back it up with a enum supporting various logging severity; approach is followed by HTTP_CLIENT and REST_CLIENT and is found useful by adjusting specific log levels while debugging EaR external KMS connectivity issues.